src/lib.rs

```diff
@@ -3,27 +3,264 @@
 #![cfg_attr(not(feature = "std"), no_std)]
 extern crate alloc;
 
-mod pointer;
-pub use pointer::*;
-mod token;
-pub use token::*;
-pub mod index;
-pub use index::Index;
-
-mod error;
-pub use error::*;
+use core::{fmt, num::ParseIntError};
 
+// TODO(chanced): make these pub?
 mod assign;
-pub use assign::*;
+pub use assign::{Assign, AssignError, Assignment};
 mod delete;
-pub use delete::*;
+pub use delete::Delete;
 mod resolve;
-pub use resolve::*;
+pub use resolve::{Resolve, ResolveError, ResolveMut};
 
 pub mod prelude;
 
+// TODO(chanced): move these into lib.rs
 mod tokens;
 pub use tokens::*;
+mod pointer;
+pub use pointer::*;
+mod token;
+pub use token::*;
+pub mod index;
+pub use index::Index;
 
 #[cfg(test)]
 mod arbitrary;
+
+/// Indicates that a `Pointer` was malformed and unable to be parsed.
+#[derive(Debug, PartialEq)]
+pub enum ParseError {
+    /// `Pointer` did not start with a backslash (`'/'`).
+    NoLeadingBackslash,
+
+    /// `Pointer` contained invalid encoding (e.g. `~` not followed by `0` or
+    /// `1`).
+    InvalidEncoding {
```

Comment:
One caveat to take into account for struct-like variants like this is that you're committing to the fields signature being part of the public interface. Probably ok here, I think.

Reply:
oh, I hadn't considered that. I think I'm okay with it in this case. Any changes to the error are going to come with other breaking changes. But yea, I'll definitely keep that in mind in the future.

Thanks.